### PR TITLE
Add support for magnum

### PR DESCRIPTION
**Pull request by mbanszex**
Change-Id: I00041cc09ecd65ea82b15c532d9dba0669659663
Signed-off-by: Martin Banszel <martinx.banszel@intel.com>

### CODE REVIEW

**Comment by rbbratta** [comment]
It seems there are several flake8 and other errors.  Are you able to find them in the travis-ci output?  It might be too verbose.

**Comment by mbanszex** [comment]
they are plenty. Seems that the settings is different from what we had in gerrit.

**Comment by rbbratta** [comment]
flake8 and pylint are passing.   We output all the flake8, pylint warning and only a certain subset are considered fatal.

I think right now it is the unittests that are failing.

**Comment by rbbratta** [comment]
All are all the unittest failures related to the pull request?  Because in this case, ensuring unittests pass, the standard procedure is to squash.  The general principal being that unittests should pass for every commit in the git history.  Or in other words, no commit in the git history should fail unittests.  This allows for git bisect and other tools to work.

If the unittest failures are in other parts of the code then, it would make sense to split those fixes into a different commits since they aren't directly related to this change.

File: taf/testlib/linux/service_lib.py
Comment (esmacgil) [inline taf/testlib/linux/service_lib.py]: These "factory" methods should be merged into their respective `__init__` methods.

File: taf/testlib/magnum.py
Comment (esmacgil) [inline taf/testlib/magnum.py]: Please include the trailing comma for literals when they are formatted in this manner.

File: taf/testlib/magnum.py
Comment (esmacgil) [inline taf/testlib/magnum.py]: When keys and values of the same dictionary are needed within a loop, please iterate over `.items()` instead of only `.keys()`.

But here a `ChainMap` would be yet better.

File: taf/testlib/rrdtool_graph.py
Comment (esmacgil) [inline taf/testlib/rrdtool_graph.py]: Please add trailing commas too.

File: taf/testlib/linux/bench.py
Comment (NazarTkachuk) [inline taf/testlib/linux/bench.py]: Incorrect license. Need to update on Apache v2.0

File: utils/ab_parser.py
Comment (NazarTkachuk) [inline utils/ab_parser.py]: Incorrect license. Need to update on Apache v2.0

File: taf/testlib/linux/etcd_helper.py
Comment (NazarTkachuk) [inline taf/testlib/linux/etcd_helper.py]: Add copyright statement

File: taf/testlib/linux/kubernetes.py
Comment (NazarTkachuk) [inline taf/testlib/linux/kubernetes.py]: Add copyright statement

File: taf/testlib/linux/utils.py
Comment (NazarTkachuk) [inline taf/testlib/linux/utils.py]: Add copyright statement

File: utils/images/onp_benchmark/Dockerfile
Comment (NazarTkachuk) [inline utils/images/onp_benchmark/Dockerfile:1]: Add line 
MAINTAINER intel.com

File: utils/images/onp_benchmark/invoke_test.py
Comment (NazarTkachuk) [inline utils/images/onp_benchmark/invoke_test.py]: Incorrect license

File: utils/images/onp_benchmark/Dockerfile
Comment (NazarTkachuk) [inline utils/images/onp_benchmark/Dockerfile]: we don't share internal domain names and ip adresses

File: utils/images/onp_benchmark/nginx.conf
Comment (NazarTkachuk) [inline utils/images/onp_benchmark/nginx.conf:16]: Add copyright statement

Comment (rbbratta) [reply utils/images/onp_benchmark/nginx.conf:16]: no copyright on config data

**Comment by rbbratta** [comment]
Fixed copyright to 2016 - 2017.  Remember when editing files to update copyright to 2017.

File: taf/testlib/linux/service_lib.py
Comment (mbanszex) [reply taf/testlib/linux/service_lib.py]: Done.

File: taf/testlib/magnum.py
Comment (mbanszex) [reply taf/testlib/magnum.py]: Done.

File: taf/plugins/pytest_sut_monitor.py
Comment (esmacgil) [inline taf/plugins/pytest_sut_monitor.py]: Call super please

File: taf/plugins/pytest_sut_monitor.py
Comment (esmacgil) [inline taf/plugins/pytest_sut_monitor.py:373]: Is it possible to change our pylint so that this isn't a stopper?  I like that pylint tells us about this, but I would prefer that it not be a blocking error.

File: taf/testlib/linux/bench.py
Comment (esmacgil) [inline taf/testlib/linux/bench.py]: This is what suppress was made for and please include the attribute name in the Attribute exception:

```
with suppress(KeyError):
  return self._config[key]
raise AttributeError(key)
```

File: taf/testlib/linux/bench.py
Comment (esmacgil) [inline taf/testlib/linux/bench.py:133]: This looks like a mapping lookup function:

```
try:
  server_file, client_file, label = MODE_MAPPING[mode]
except KeyError:
  raise Exception(...)

set_node_selector(server_file, label)
set_node_selector(client_file, label)
```

File: taf/testlib/linux/bench.py
Comment (esmacgil) [inline taf/testlib/linux/bench.py]: Please define a constant for these String literals.

File: taf/testlib/linux/bench.py
Comment (esmacgil) [inline taf/testlib/linux/bench.py]: Since we aren't unifying the return statements, I'd prefer this to be the short path:

```
if self._test_type is not None:
  return self._test_type
client_file = ...
...
```

thus allowing the de-indentation of the body of this function.

**Comment by rbbratta** [comment]
Needs review

File: taf/testlib/linux/etcd_helper.py
Comment (esmacgil) [inline taf/testlib/linux/etcd_helper.py]: Please use suppress:

```
with suppress(AttributeError):
  return self._latest_id
with suppress(Exception):
  for _ in range(2):
    with suppress(etcd.EtcdKeyNotFound):
      self._latest_id = ...
      return self._latest_id
    self.init_etcd()
raise EtcdHelperException(...)
```

File: taf/testlib/rrdtool_graph.py
Comment (esmacgil) [inline taf/testlib/rrdtool_graph.py]: This should have a comma, did flake8-comma not flag this?

File: taf/testlib/rrdtool_graph.py
Comment (esmacgil) [inline taf/testlib/rrdtool_graph.py]: Another missing comma

File: taf/testlib/rrdtool_graph.py
Comment (esmacgil) [inline taf/testlib/rrdtool_graph.py]: Missing comma

File: taf/plugins/pytest_sut_monitor.py
Comment (rbbratta) [reply taf/plugins/pytest_sut_monitor.py:373]: add it here https://github.com/taf3/taf/blob/ci/project_checker.py#L578

Comment (rbbratta) [reply taf/plugins/pytest_sut_monitor.py:373]: err it's not in the fatal list, so pylint won't block on W0613.  So it is acting just as a warning currently.

https://github.com/taf3/taf/blob/ci/project_checker.py#L958

File: taf/testlib/linux/etcd_helper.py
Comment (rbbratta) [reply taf/testlib/linux/etcd_helper.py]: why are we suppressing Exception?  I thought we implemented something that had an exclude list so we don't suppress SyntaxError, NameError, etc.  

No code should ever catch just Exception.

File: taf/plugins/pytest_sut_monitor.py
Comment (mbanszex) [reply taf/plugins/pytest_sut_monitor.py]: Done

File: taf/testlib/linux/bench.py
Comment (mbanszex) [reply taf/testlib/linux/bench.py]: Done.

File: taf/testlib/linux/etcd_helper.py
Comment (mbanszex) [reply taf/testlib/linux/etcd_helper.py]: I used the Mac's suppress version leaving out the supress(Exception). If something unknown happens it should drop down to the caller and the exception call stack should not be overwritten with calling raise.

**Comment by rbbratta** [comment]
Review scrub: needs review

**Comment by rbbratta** [comment]
Mac to review again

File: taf/testlib/linux/bench.py
Comment (esmacgil) [inline taf/testlib/linux/bench.py]: Is this correct?  Am I reading this correctly: that if the first `test_type` in `self.PARSERS` is not in `client_file` then the exception will be raised.  Shouldn't the `raise` statement be at the same indentation as the `for` statement?

Also, since this code doesn't need the value at `client_file[test_type]`, we can test for the presence instead of doing the look up:

```
for test_type in self.PARSERS:
    if test_type in client_file:
        self._test_type = test_type
        return self._test_type
raise BenchmarkException(...)
```

or

```
with suppress(StopIteration):
    test_type = next(test_type for test_type in self.PARSERS if test_type in client_file)
    self._test_type = test_type
    return self._test_type
raise BenchmarkException(...)
```

File: taf/testlib/tempest_clients/magnum/clients/cluster_client.py
Comment (esmacgil) [inline taf/testlib/tempest_clients/magnum/clients/cluster_client.py:126]: Do not use modulo (`%`) when calling logger methods.

`self.LOG.error('Cluster %s entered an exception state', cluster_id)`

**Review by esmacgil — CHANGES_REQUESTED** [review]

File: taf/testlib/tempest_clients/magnum/clients/cluster_client.py
Comment (mbanszex) [reply taf/testlib/tempest_clients/magnum/clients/cluster_client.py:126]: This magnum client code is taken over from magnum project. I wanted to keep it in the original shape. Except for some imports adjustments, I haven't touched it.

File: taf/testlib/linux/bench.py
Comment (mbanszex) [reply taf/testlib/linux/bench.py]: yes, that's a bug.

File: taf/testlib/virtual_env.py
Comment (esmacgil) [inline taf/testlib/virtual_env.py]: Logging with format.

File: taf/testlib/virtual_env.py
Comment (esmacgil) [inline taf/testlib/virtual_env.py]: Won't the next statement raise some exception if `public_router` is not a dictionary or has no values?

The only thing this statement seems to protect against , beyond the statement that follows it, is an _empty_ `defaultdict`, while it would pass through many other invalid values: `True`, `1`, `[0]`, `object()` and many more.

File: taf/testlib/virtual_env.py
Comment (esmacgil) [inline taf/testlib/virtual_env.py]: Please return the result of the Boolean expression instead of testing the Boolean expression and returning a Boolean literal.

```
return _net_cfg.public_network_id and (...)
```

**Review by esmacgil — CHANGES_REQUESTED** [review]

File: taf/testlib/virtual_env.py
Comment (mbanszex) [reply taf/testlib/virtual_env.py]: Yes, that's right. This assert is meant only to catch the situation when the cluster creation fails and the tempest returns None. We could test a lot of things to make sure that we got a correct dictionary. But if it woudn't have e.g. the id key, something would be very wrong and then I think the KeyError exception or similar is appropriate.

File: taf/testlib/virtual_env.py
Comment (mbanszex) [reply taf/testlib/virtual_env.py]: Done

File: taf/testlib/virtual_env.py
Comment (mbanszex) [reply taf/testlib/virtual_env.py]: Fixed

**Review by esmacgil — APPROVED** [review]

**Comment by esmacgil** [comment]
I see you say that comments have been addressed, but where is the latest change set?

**Comment by mbanszex** [comment]
Sorry, I somehow lost part of the changes. Added now. Before merging to master, we need to squash all the commits either into one or refactor the commits so that they make some sense. Currently they just reflect the review process, not the logical changes.

File: taf/testlib/virtual_env.py
Comment (esmacgil) [inline taf/testlib/virtual_env.py]: Looking at this again, would it make more sense to put the `not create_external_router` clause before `_net_cfg.public_router_id`?  Because the latter could raise an AttributeError even when the result should be True.

```return _net_cfg.public_network_id and (not create_external_router or _net_cfg.public_router_id)```

Comment (mbanszex) [reply taf/testlib/virtual_env.py]: Done

File: taf/plugins/pytest_sut_monitor.py
Comment (esmacgil) [inline taf/plugins/pytest_sut_monitor.py:233]: can we make this command a list instead of a string?

```
folders = self.exec_command(['find', '{}/*'.format(collectd_folder), '-maxdepth', '1', '-type', 'd', '-print0']).split('\0')
```

File: taf/plugins/pytest_sut_monitor.py
Comment (esmacgil) [reply taf/plugins/pytest_sut_monitor.py]: Why don't I see a call to super?

File: taf/testlib/linux/service_lib.py
Comment (esmacgil) [inline taf/testlib/linux/service_lib.py:229]: is there a reason to do this join when the command execution function should do it for us?

File: taf/testlib/linux/service_lib.py
Comment (esmacgil) [inline taf/testlib/linux/service_lib.py:200]: is there a reason to do this join when the command execution function should do it for us?

File: taf/plugins/pytest_sut_monitor.py
Comment (mbanszex) [reply taf/plugins/pytest_sut_monitor.py:233]: LinuxHostBash.cli_send_command() and paramiko exec_command() accept a string, not a list of strings.
List might be better but I don't want to change this in the whole TAF in this PR. We can change that in a separate PR.

File: taf/testlib/linux/service_lib.py
Comment (mbanszex) [reply taf/testlib/linux/service_lib.py:200]: The run_func is given by the caller. At the point all exec_command or cli_send_command implentations in TAF expect the command to be a string. Sometimes thy even call command.split(' ') -- which is totaly wrong. (taf/testlib/clinns.py:174)
This should be fixed globaly in a separate PR.

**Review by esmacgil — APPROVED** [review]

**Pull request merged**
